File: problem189.cpp
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size(); // size of n
        // if k is greater than size of nums, we need remainder of n / k
        k =  k % n; // new k
        // Base case
        if(k == 0){
            return;
        }

        int count = 0; // number of elements are changed
        for(int i =0; i < n; i++){
            int current = i; // this is starting point 
            int prev = nums[current]; // this is the number will be changed 
            if(count >= n){
                break;
            }
            bool run = true;
            while(run){
                int nextIndex = (current + k) % n; // next index
                int save = nums[nextIndex]; // save next index number
                nums[nextIndex] = prev; // nextindex will have prev number 
                prev = save; // prev number will be updated to next number
                current = nextIndex; // also current index will be updated too.
                count++; // number of changed element in array will be increasd by one
                if(current == i){
                    run = false;
                }
            }   
        }
    }
};

File: problem189.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # Base case
        if k % len(nums) == 0:
            return nums
        # reverse
        
        # if k is greater than n
        k = k % len(nums)

        def reverse(start, end):
            while start < end:
                temp = nums[start]
                nums[start] = nums[end]
                nums[end] = temp
                start += 1
                end -= 1
        
        # Step one reverse whole array
        reverse(0, len(nums)-1)
        print(nums)

        # Step2: reverse 0 to k
        reverse(0, k-1)
        print(nums)

        #Step3 : reverse  k to n
        reverse(k, len(nums)-1)
        print(nums)

